Documentation: warning on multi-threading/processing (#5287)
pyodide/pyodide: docs/development/new-packages.md

{func}`micropip.install` if they have a pure Python wheel. Check if this is the
case by trying `micropip.install("package-name")`.

> Due to [current limitations of webassemly](https://pyodide.org/en/stable/usage/wasm-constraints.html),

Reviewer: ```suggestion
Due to [current limitations of WebAssembly](https://pyodide.org/en/stable/usage/wasm-constraints.html),
```

packages that make heavy use of multi-threading and multi-processing **may not** work
without explicit support from maintainers. A workaround could be to avoid creating multiple

> threads when on wasm using a check like:

Reviewer: ```suggestion
threads when on WASM using a check like:
```

> Due to [current limitations of WebAssembly](https://pyodide.org/en/stable/usage/wasm-constraints.html),
> packages that make heavy use of multi-threading and multi-processing **may not** work
> without explicit support from maintainers. A workaround could be to avoid creating multiple
> threads when on WASM using a check like:

Reviewer 2: Maybe:
```suggestion
Because Pyodide [does not support threading or multiprocessing](https://pyodide.org/en/stable/usage/wasm-constraints.html),
packages that use threading or multiprocessing will not work without a patch to disable it. For example,
```

```py

> is_wasm = sys.platform == "emscripten" or platform.machine() in ["wasm32", "wasm64"]

Reviewer 2: It's always a bit annoying deciding how best to check this. `sys.platform == "emscripten"` implies `platform.machine() in ["wasm32", "wasm64"]` so it's a bit redundant. The Python test suite does:
```py
def _can_start_thread() -> bool:
    if sys.platform == "emscripten":
        return sys._emscripten_info.pthreads
    return platform.machine() not in ("wasm32", "wasm64")

can_start_thread = _can_start_thread()
```
it would be nice if threading/multiprocessing offered some API to query whether they work on the platform...

Author: I agree. I don't know what `sys._emscripten_info.pthreads` returns as if i understand threads are supported but due to spectre they were disabled. If I try to compile `argon2-cffi` with `-lpthread` it gives no issue, but the error persists.

I like your function more than mine, but maybe we should just remove the "suggested workaround" to not make any wrong assumptions. What do you think?

if is_wasm:

>   # set n_threads = 1

Reviewer 2: Since this is pseudocode anyways maybe we should just say:
```suggestion
  n_threads = 1
```

{func}`micropip.install` if they have a pure Python wheel. Check if this is the
case by trying `micropip.install("package-name")`.

Because Pyodide [does not support threading or multiprocessing](https://pyodide.org/en/stable/usage/wasm-constraints.html),
packages that use threading or multiprocessing will not work without a patch to disable it. For example,
the following snippet will determine if the platform supports creating new threads.

```py
def _can_start_thread() -> bool:
    if sys.platform == "emscripten":
        return sys._emscripten_info.pthreads
    return platform.machine() not in ("wasm32", "wasm64")

can_start_thread = _can_start_thread()

if not can_start_thread:

>   # set n_threads = 1

Reviewer 2: ```suggestion
  n_threads = 1
```
